### PR TITLE
Confidence classes of SVs (Manta and BIC-Seq2)

### PR DESCRIPTION
**Pull request by dmaziec**
This PR introduces:
1. A new step that infers confidence classes of germline SVs (Manta)

- The step is executed  in the downstream SV germline pipeline after the `sansa/VEP annotation` step.
- It requires an output VCF file containing results from Manta as an input.
- Possible class confidence are: LOW, MEDIUM, HIGH, NA. We assign the confidence class for each sample in the VCF file under a new `FORMAT` field: `CF`. 
- The values are calculated based on the length and the number of split and spanning reads for the given variant according to the following formulas: 

```
if variant == INSERTION, then class = NA, 

else if ( (length > 250bp or svtype == TRANSLOCATION) and split_reads >= 5 and prop_split_reads >= 0.3 and spanning_reads >=b5 and prop_spanning_reads >= 0.3), then class = HIGH 

else if ((length <= 250bp  and svtype != TRANSLOCATION ) and split_reads > 5 and prop_split_reads > 0.3), then class = HIGH 

else if ((length > 250bp or variant == TRANSLOCATION) and split_reads >= 3 and prop_split_reads >= 0.3 and spanning_reads >= 3 and prop_spanning_reads >= 0.3), then class = MEDIUM

else if ( (length <= 250bp and svtype != TRANSLOCATION)  and split_reads > 3 and prop_split_reads > 0.3), then class = MEDIUM

else class = LOW 
```
where: 

```
length - the absolute value of the SVLEN field
split_reads - the number of alternative split reads from the SR field
svtype - type of structural variant
spanning_reads - the number of alternative spanning reads from the PR field
prop_split_reads  =  alternative split reads / (alternative split reads + reference split reads) #from the SR field
prop_spanning_reads  = alternative spanning reads / (alternative spanning reads + reference spanning reads)  #from the PR field
```

Currently, we do not assign confidence classes for insertions, thus these variants have assigned `NA`. 

2. A new step that infers confidence classes of germline CNVs (BIC-Seq2)
- The step is executed  in the downstream CNV germline pipeline after the `sansa/VEP annotation` step.
- It requires an output VCF file containing results from BIC-Seq2 as an input.
- Possible class confidence are: LOW,  HIGH. We assign the confidence class for each sample in the VCF file under a new `FORMAT` field: `CF`. 
- The values are calculated based on the length and log2_copyRatio provided by BIC-Seq2 using the following formulas: 

```
if length > 1Mb and (log2_copyRatio > 0.4 or log2_copyRatio < -0.8), then class = HIGH 
else class = LOW  
```

### CODE REVIEW

File: cwl/BICseq2_add_confidence.cwl [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    inputBinding:
+      prefix: -i
+    doc: expect the path to the sample vcf gz file
+
+  - id: output_vcf
+    type: string
+    default: "output.vcf"
+    inputBinding:
+      prefix: -o
+    doc: base name of output vcf gz file
```
Comment (B3rse) [inline cwl/BICseq2_add_confidence.cwl]: base name of output vcf gz file -> base name of output vcf file

File: cwl/manta_add_confidence.cwl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    inputBinding:
+      prefix: -i
+    doc: expect the path to the sample vcf gz file
+
+  - id: output_vcf
+    type: string
+    default: "output.vcf"
+    inputBinding:
+      prefix: -o
+    doc: base name of output vcf gz file
```
Comment (B3rse) [inline cwl/manta_add_confidence.cwl]: same as ^

File: portal_objects/workflows/workflow_BICseq2_add_confidence.json [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    "arguments": [
+        {
+            "argument_format": "vcf",
+            "argument_type": "Input file",
+            "workflow_argument_name": "input_vcf"
+        },
+        {
+            "argument_format": "vcf_gz",
+            "argument_type": "Output processed file",
+            "workflow_argument_name": "vcf_sv_confidence"
```
Comment (B3rse) [inline portal_objects/workflows/workflow_BICseq2_add_confidence.json]: here we need to specify the secondary files as well, you can probably copy it from some other workflow

File: portal_objects/workflows/workflow_manta_add_confidence.json [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    "arguments": [
+        {
+            "argument_format": "vcf",
+            "argument_type": "Input file",
+            "workflow_argument_name": "input_vcf"
+        },
+        {
+            "argument_format": "vcf_gz",
+            "argument_type": "Output processed file",
+            "workflow_argument_name": "vcf_sv_confidence"
```
Comment (B3rse) [inline portal_objects/workflows/workflow_manta_add_confidence.json]: here we need to specify the secondary files as well, you can probably copy it from some other workflow

File: portal_objects/workflows/workflow_BICseq2_add_confidence.json (hunk 3 above)
Comment (B3rse) [reply portal_objects/workflows/workflow_BICseq2_add_confidence.json]: or does it work without? like if you don't specify the secondary files it is still getting the index file out?

File: portal_objects/workflows/workflow_BICseq2_add_confidence.json [hunk 5]
```diff
@@ -0,0 +1,74 @@
+{
+    "accession": "GAPWFR7F3KN3",
+    "app_name": "workflow_BICseq2_add_confidence",
+    "app_version": "VERSION",
+    "arguments": [
+        {
+            "argument_format": "vcf",
```
Comment (B3rse) [inline portal_objects/workflows/workflow_BICseq2_add_confidence.json]: should this be vcf_gz? again does it work fine also if it is only vcf?

File: portal_objects/workflows/workflow_manta_add_confidence.json [hunk 6]
```diff
@@ -0,0 +1,74 @@
+{
+    "accession": "GAPWFR8V3KN3",
+    "app_name": "workflow_manta_add_confidence",
+    "app_version": "VERSION",
+    "arguments": [
+        {
+            "argument_format": "vcf",
```
Comment (B3rse) [inline portal_objects/workflows/workflow_manta_add_confidence.json]: vcf_gz

File: portal_objects/metaworkflows/SV_germline.json [hunk 7]
```diff
@@ -217,6 +217,29 @@
...
+        "behavior_on_capacity_limit": "wait_and_retry"
+      },
+      "input": [
+        {
+          "argument_name": "input_vcf",
+          "argument_type": "file",
+          "source": "workflow_annotateSV_sansa_vep_vcf-check",
+          "source_argument_name": "annotated_SV_vcf"
+        }
+      ]
```
Comment (B3rse) [inline portal_objects/metaworkflows/SV_germline.json:246]: we need to add 

      "custom_pf_fields": {
          "vcf_sv_confidence": {
              "variant_type": "SV",
              "file_type": "confidence vcf",
              "description": "intermediate file"
          }

File: portal_objects/metaworkflows/cnv_germline.json [hunk 8]
```diff
@@ -223,6 +223,26 @@
...
+      "name": "workflow_BICseq2_add_confidence",
+      "workflow": "dcee73e2-66b2-4f9c-8bd7-43243b7d84fe",
+      "config": {
+        "instance_type": "t3.small",
+        "ebs_size": "2x",
+        "EBS_optimized": true,
+        "spot_instance": true,
+        "run_name": "run_workflow_BICseq2_add_confidence",
+        "behavior_on_capacity_limit": "wait_and_retry"
+      },
```
Comment (B3rse) [inline portal_objects/metaworkflows/cnv_germline.json:236]: we need to add 

      "custom_pf_fields": {
          "vcf_sv_confidence": {
              "variant_type": "CNV",
              "file_type": "confidence vcf",
              "description": "intermediate file"
          }

Comment (B3rse) [reply portal_objects/metaworkflows/cnv_germline.json:236]: this fields are used by tibanna to tag the file on the portal, we will probably improve this as well and have more structured tags but for now we need to stick with this

**Review by B3rse — CHANGES_REQUESTED** [review]
comments

File: dockerfiles/sv_germline_granite/scripts/SV_confidence.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        Variant class: a single variant record with the calculated confidence classes
+
+    """
+    svtype = vnt_obj.get_tag_value("SVTYPE")
+
+    # difference in length between REF and ALT alleles
+    svlen = 0
+    # the length condition is not applicable to translocations and insertions
+    if svtype not in [TRANSLOCATION_SVTYPE_MANTA, INSERTION_SVTYPE_MANTA]:
+        svlen = abs(int(vnt_obj.get_tag_value("SVLEN")))
```
Comment (dominikglodzikhms) [inline dockerfiles/sv_germline_granite/scripts/SV_confidence.py:95]: If the 'SVLEN' was empty or NA for any reason (we don't understand), what would your algorithm do? I suggest adding an explicit way of handling it, even if we haven't seen it yet.

File: dockerfiles/sv_germline_granite/scripts/test/test_SV_confidence.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    b = os.popen("bgzip -c -d test/files/confidence_classes_manta_out.vcf.gz")
+
+    # Test
+    assert [row for row in a.read()] == [row for row in b.read()]
+
+
+def test_manta_multiple_gentype_ids(tmp_path):
+    # Variables and Run
+    args = {
+        "input": "test/files/confidence_classes_manta_3_genotype_ids_in.vcf.gz",
```
Comment (dominikglodzikhms) [inline dockerfiles/sv_germline_granite/scripts/test/test_SV_confidence.py:41]: Great that you thought about careful tests. Could you explain what is different in 'test_manta_multiple_gentype_ids' compared to 'test_manta' and the test files?

File: dockerfiles/sv_germline_granite/scripts/SV_confidence.py (hunk 9 above)
Comment (dmaziec) [reply dockerfiles/sv_germline_granite/scripts/SV_confidence.py:95]: If  there is something wrong with `SVLEN` (e.g. the value is missing or not provided) , the `get_tag_value` function will raise an error and stop the execution. The error will be the following: `ERROR in variant INFO field, SVLEN tag is in the wrong format.` 

Here is the implementation of this function: https://github.com/dbmi-bgm/granite/blob/e6898a090b92846678277f2c972426e5078a6761/granite/lib/vcf_parser.py#L268 

 If there is `NA`, we won't be able to convert it to an integer so the `int` method will raise an error as well.

File: dockerfiles/sv_germline_granite/scripts/test/test_SV_confidence.py (hunk 10 above)
Comment (dmaziec) [reply dockerfiles/sv_germline_granite/scripts/test/test_SV_confidence.py:41]: In `test_manta` we test a VCF file that stores various combinations of possible variants classes and there is only one sample tested. In `test_manta_multiple_gentype_ids` we test a VCF file that stores several variants for three samples to check if we can evaluate confidence classes for all the samples provided in the file. The same approach is applied for the BIC-Seq2 tests. 

I will leave comments on this in the script to make it clearer.

Comment (dominikglodzikhms) [reply dockerfiles/sv_germline_granite/scripts/test/test_SV_confidence.py:41]: In` test_manta_multiple_gentype_ids `, does it mean that each of 3 samples will get its own confidence class for the same variant?

Comment (dmaziec) [reply dockerfiles/sv_germline_granite/scripts/test/test_SV_confidence.py:41]: Yes. Example: 
```
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	NA24385_sample-WGS	NA24143_sample-WGS	NA24149_sample-WGS
chr1	5112477	MantaBND:16:48380:48381:1:0:0:1	G	G]chr4:150658967]	15	MinQUAL;SampleFT	SVTYPE=BND;MATEID=MantaBND:16:48380:48381:1:0:0:0;IMPRECISE;CIPOS=-550,550;BND_DEPTH=130;MATE_BND_DEPTH=29;CSQ=G]chr4|intergenic_variant|MODIFIER|||||||||||||||||||||||	GT:FT:GQ:PL:PR:CF	0/0:HomRef;MinGQ:7:44,0,70:5,3:LOW	0/0:HomRef:53:0,3,172:11,2:LOW	0/1:MinGQ:6:63,0,3:1,5:LOW
```

File: portal_objects/workflows/workflow_BICseq2_add_confidence.json [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    "institution": "INSTITUTION_UUID",
+    "name": "workflow_BICseq2_add_confidence_VERSION",
+    "aliases": ["cgap:workflow_BICseq2_add_confidence_VERSION"],
+    "steps": [
+        {
+            "inputs": [
+                {
+                    "meta": {
+                        "cardinality": "single",
+                        "file_format": "vcf",
```
Comment (B3rse) [inline portal_objects/workflows/workflow_BICseq2_add_confidence.json]: that should be vcf_gz as well, same in the other

File: portal_objects/workflows/workflow_manta_add_confidence.json [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    "institution": "INSTITUTION_UUID",
+    "name": "workflow_manta_add_confidence_VERSION",
+    "aliases": ["cgap:workflow_manta_add_confidence_VERSION"],
+    "steps": [
+        {
+            "inputs": [
+                {
+                    "meta": {
+                        "cardinality": "single",
+                        "file_format": "vcf",
```
Comment (B3rse) [inline portal_objects/workflows/workflow_manta_add_confidence.json]: also this one

**Review by B3rse — APPROVED** [review]

**Pull request merged**